### PR TITLE
Update looting

### PR DESCRIPTION
**Pull request by moseleyc**
## Description
-Add item looting, and few minor fixes

Note: ImportLoot functionality not added yet, but place holder added

## Checklist
- [x] Tested and working locally
- [ ] Reviewed
- [ ] Tested and working on dev server
- [ ] Deployed to staging

## How to QA
- [x] Kill creature with one loot-table, item is dropped (based on chance value)
- [x] Kill creature with one loot-table, gold is dropped (based on chance value)
- [x] Kill creature with multiple loot-tables, item(s) are dropped (based on chance value & loot-table)
- [x] Kill creature with multiple loot-tables, gold(s) are dropped (based on chance value & loot-table)
- [x] Kill creature with no loot-table, gold is dropped (based on loot, not loot-table)

- [x] User kills creature, able to pick up items
- [x] User kills creature, otherUser unable to pick up items until deathpile duration exceeded
- [x] User kills creature, grpedUser able to pick up items

- [x] User with damage, able to be healed
- [x] User able to heal damaged otherUser
- [x] User able to attack Monster with Spell

## Impacted Areas in Hybrasyl

-Monster (added Private Class MonsterLoot / added private variable to hold the spawnsLoot info / added Monster DeathPile Assignment / fixed bug with LootableXP calculation)
-Creature (fixed issue with healing / added DeathPileLooting assignment (for mob loot) / adjusted deadMobs to use Monster.OnDeath)
-World (added logic to OnPickup to remove DeathPile assignment if the item is picked up, and to update (avoid items appearing on the ground after already picked up))
-User (fixed bug with casting cost)
-VisibleObjects (fixed bug with DeathPileTime evaluation)
-Castables.Designer (removed the assignment of Damage / Heal, was causing issue with UseCastable Null evaluation)

### CODE REVIEW

File: hybrasyl/Objects/Monster.cs [hunk 1]
```diff
@@ -42,6 +42,8 @@ public class Monster : Creature, ICloneable
 
         private Spawn _spawn;
 
+        private MonsterLoot spawnLoot;
```
Comment (norrismiv) [inline hybrasyl/Objects/Monster.cs]: private variables should always begin with an underscore. '_spawnLoot'

File: HybrasylXML/XSD/Castables.Designer.cs [hunk 2]
```diff
@@ -1367,8 +1367,6 @@ public Effects()
         {
             this._statuses = new Statuses();
             this._statModifiers = new StatModifiers();
-            this._damage = new Damage();
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.Designer.cs]: Why were these removed?

File: hybrasyl/Objects/Creature.cs [hunk 3]
```diff
@@ -302,8 +302,8 @@ public virtual List<Creature> GetTargets(Castable castable, Creature target = nu
                         Logger.Error($"GetTargets: {castable.Name} - intent was for exact clicked target but no target was passed?");
                     else
                         // Heal spels can be cast on players, other spells can be cast on attackable creatures
-                        if ((!castable.Effects.Damage.IsEmpty && target.Condition.IsAttackable) ||
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: IsEmpty should be the proper pattern. Damage should never be null.

File: hybrasyl/Objects/Creature.cs [hunk 4]
```diff
@@ -302,8 +302,8 @@ public virtual List<Creature> GetTargets(Castable castable, Creature target = nu
...
-                        (castable.Effects.Damage.IsEmpty && target is User))
+                        if ((castable.Effects.Damage != null && target.Condition.IsAttackable) ||
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: Damage should never be null. This is likely caused by the change made to The castables designer file. See comment regarding that.

File: hybrasyl/Objects/Creature.cs [hunk 5]
```diff
@@ -908,7 +913,11 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
                 }
             }
             // Now flood away
-            foreach (var dead in deadMobs) dead.OnDeath();
+            foreach (var dead in deadMobs)
+            {
+                if (dead is Monster) (dead as Monster).OnDeath();
+                else dead.OnDeath();
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: this if/else statement does exactly the same thing. There is no need to cast dead as a monster.

File: hybrasyl/Objects/Monster.cs [hunk 6]
```diff
@@ -310,6 +324,89 @@ public object Clone()
         {
             return this.MemberwiseClone();
         }
-    }
 
+        private class MonsterLoot
```
Comment (norrismiv) [inline hybrasyl/Objects/Monster.cs]: This entire class looks like it should be static. Spawn related static objects are contained within the monolith class, which is designed to handle methods like this.

File: hybrasyl/Objects/User.cs [hunk 7]
```diff
@@ -1219,7 +1219,7 @@ public bool ProcessCastingCost(Castable castable)
 
             // Check that all requirements are met first. Note that a spell cannot be cast if its HP cost would result
             // in the caster's HP being reduced to zero.
-            if (reduceHp > Hp || reduceMp >= Mp || costs.Gold > Gold || (removeNumItems != 0 && costs.Items.Count == removeNumItems)) return false;
+            if (reduceHp > Hp || reduceMp >= Mp || costs.Gold > Gold || (costs.Items != null && costs.Items.Count != removeNumItems)) return false;
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: this looks like it's related to casting, and should be created in a separate PR.

File: hybrasyl/Objects/VisibleObject.cs [hunk 8]
```diff
@@ -82,8 +82,8 @@ public bool CanBeLooted(string username, out string error)
             if (DeathPileTime == null) return true;
             if (DeathPileOwner == username) return true;
             if (DeathPileAllowedLooters.Contains(username) &&
-                (DateTime.Now - DeathPileTime.Value).Seconds > Constants.DEATHPILE_GROUP_TIMEOUT) return true;
-            if ((DateTime.Now - DeathPileTime.Value).Seconds > Constants.DEATHPILE_RANDO_TIMEOUT) return true;
+                (DateTime.Now - DeathPileTime.Value).TotalSeconds > Constants.DEATHPILE_GROUP_TIMEOUT) return true;
```
Comment (norrismiv) [inline hybrasyl/Objects/VisibleObject.cs]: Is this related to looting items from mobs? If so, deathpiletime should not be used. New functionality should be created to handle looting versus death pile of a user.

File: hybrasyl/World.cs [hunk 9]
```diff
@@ -1407,6 +1407,8 @@ private void PacketHandler_0x07_PickupItem(Object obj, ClientPacket packet)
 
                     Logger.DebugFormat("Removing {0}, qty {1} from {2}@{3},{4}",
                         item.Name, item.Count, user.Map.Name, x, y);
+                    item.DeathPileTime = null;
```
Comment (norrismiv) [inline hybrasyl/World.cs]: this seems unnecessary as the item is removed from the map directly below.

File: hybrasyl/World.cs [hunk 10]
```diff
@@ -1416,13 +1418,17 @@ private void PacketHandler_0x07_PickupItem(Object obj, ClientPacket packet)
                         user.Map.Insert(item, user.X, user.Y);
                         user.SendMessage(string.Format("You can't carry any more {0}.", item.Name), 3);
                     }
+                    Update();
                 }
                 else
                 {
                     Logger.DebugFormat("Removing {0}, qty {1} from {2}@{3},{4}",
                         item.Name, item.Count, user.Map.Name, x, y);
+                    item.DeathPileTime = null;
```
Comment (norrismiv) [inline hybrasyl/World.cs]: see comment about being unnecessary above.

File: hybrasyl/Objects/Creature.cs [hunk 11]
```diff
@@ -820,10 +820,7 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
             if (!Condition.CastingAllowed) return false;
             if (this is User) ActivityLogger.Info($"UseCastable: {Name} begin casting {castObject.Name} on target: {target?.Name ?? "no target"} CastingAllowed: {Condition.CastingAllowed}");
 
-            var damage = castObject.Effects.Damage;
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: not sure why this was removed.

File: hybrasyl/Objects/Creature.cs [hunk 12]
```diff
@@ -302,8 +302,8 @@ public virtual List<Creature> GetTargets(Castable castable, Creature target = nu
                         Logger.Error($"GetTargets: {castable.Name} - intent was for exact clicked target but no target was passed?");
                     else
                         // Heal spels can be cast on players, other spells can be cast on attackable creatures
-                        if ((!castable.Effects.Damage.IsEmpty && target.Condition.IsAttackable) ||
-                        (castable.Effects.Damage.IsEmpty && target is User))
+                        if ((castable.Effects.Damage != null && target.Condition.IsAttackable) ||
+                        (castable.Effects.Heal != null && target is User))
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: Heal should never be null. This looks like it was caused by a change made to your castables class.

File: HybrasylXML/XSD/Castables.Designer.cs [hunk 13]
```diff
@@ -1367,8 +1367,6 @@ public Effects()
...
-            this._heal = new Heal();
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.Designer.cs]: Why was this removed?

File: hybrasyl/Objects/Creature.cs [hunk 14]
```diff
@@ -854,6 +851,14 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
                 }
                 if (castObject.Effects?.Damage != null)
                 {
+                    if (tar.DeathPileAllowedLooters.Count != 0 && !(tar.DeathPileAllowedLooters.Contains(Name))) continue;
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: not sure why this check is needed for casting.

File: hybrasyl/Objects/Creature.cs [hunk 15]
```diff
@@ -854,6 +851,14 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
...
+                    if (this is User && !tar.DeathPileAllowedLooters.Contains(Name))
+                    {
+                        var theUser = this as User;
+                        if (theUser.Grouped) tar.DeathPileAllowedLooters = theUser.Group.Members.Select(user => user.Name).ToList();
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: DeathPileAllowedLooters should not be used for Item drops from mobs. New functionality should be created to handle that.

DeathPileAllowedLooters should be used for user death piles.

File: hybrasyl/World.cs (hunk 9 above)
Comment (moseleyc) [reply hybrasyl/World.cs]: In my testing, the DeathPileTime / Ownership continued after an item was picked up. If the user set it down after, these values continued.

File: hybrasyl/Objects/VisibleObject.cs (hunk 8 above)
Comment (moseleyc) [reply hybrasyl/Objects/VisibleObject.cs]: Okay. Will look into that.

File: hybrasyl/Objects/Monster.cs (hunk 1 above)
Comment (moseleyc) [reply hybrasyl/Objects/Monster.cs]: Understood, will adjust.

File: hybrasyl/Objects/User.cs (hunk 7 above)
Comment (moseleyc) [reply hybrasyl/Objects/User.cs]: Will do.

File: HybrasylXML/XSD/Castables.Designer.cs (hunk 2 above)
Comment (moseleyc) [reply HybrasylXML/XSD/Castables.Designer.cs]: This conflicts with the way UseCastable was constructed. It evaluates Damage based on whether it's null or not. If it's null, it will then look at Heal and do the same thing. With these being set up here, Damage will never be null, so this evaluation fails. This is also related to the casting piece, so I will move these to a different PR.

File: hybrasyl/Objects/Creature.cs (hunk 11 above)
Comment (moseleyc) [reply hybrasyl/Objects/Creature.cs]: The variable doesn't do anything.

**Comment by moseleyc** [comment]
Will close and re-do. Need to separate the casting piece from the looting.
